bindings_ffi/src/v2.rs

```diff
@@ -279,46 +283,85 @@ impl From<FfiV2SubscribeRequest> for SubscribeRequest {
+#[uniffi::export(callback_interface)]
+pub trait FfiV2SubscriptionCallback: Send + Sync {
+    fn on_message(&self, message: FfiEnvelope);
 }
 
-impl From<xmtp_api_grpc::grpc_api_helper::GrpcMutableSubscription> for FfiV2Subscription {
-    fn from(subscription: xmtp_api_grpc::grpc_api_helper::GrpcMutableSubscription) -> Self {
-        Self {
-            inner_subscription: Arc::new(futures::lock::Mutex::new(subscription)),
-        }
-    }
+#[derive(uniffi::Object)]
+pub struct FfiV2Subscription {
+    tx: mpsc::Sender<FfiV2SubscribeRequest>,
+    abort: AbortHandle,
+    // we require Arc<Mutex<>> here because uniffi doesn't like &mut or self in impl
+    #[allow(clippy::type_complexity)]
+    handle: Arc<Mutex<Option<JoinHandle<Result<(), GenericError>>>>>,
 }
 
 #[uniffi::export(async_runtime = "tokio")]
 impl FfiV2Subscription {
-    pub async fn next(&self) -> Result<FfiEnvelope, GenericError> {
-        let mut sub = self.inner_subscription.lock().await;
-
-        let result = sub.next().await;
-        match result {
-            Some(Ok(envelope)) => Ok(envelope.into()),
-            Some(Err(err)) => Err(GenericError::Generic {
-                err: err.to_string(),
-            }),
-            None => Err(GenericError::Generic {
-                err: "stream closed".to_string(),
-            }),
+    pub async fn end(&self) -> Result<(), GenericError> {
+        if self.abort.is_finished() {
+            return Ok(());
+        }
+
+        let mut handle = self.handle.lock().await;
+        let handle = handle.take();
+        if let Some(h) = handle {
+            h.abort();
+            h.await.map_err(|_| GenericError::Generic {
+                err: "subscription event loop join error".into(),
+            })??;
         }
+        Ok(())
+    }
+
+    pub fn is_closed(&self) -> bool {
+        self.abort.is_finished()
     }
 
     pub async fn update(&self, req: FfiV2SubscribeRequest) -> Result<(), GenericError> {
-        let mut sub = self.inner_subscription.lock().await;
-        sub.update(req.into()).await?;
+        self.tx.send(req).await.map_err(|_| GenericError::Generic {
+            err: "stream closed".into(),
+        })?;
         Ok(())
     }
+}
 
-    pub async fn end(&self) {
-        let sub = self.inner_subscription.lock().await;
-        sub.close();
+impl FfiV2Subscription {
+    pub async fn subscribe(
+        mut subscription: GrpcMutableSubscription,
+        callback: Box<dyn FfiV2SubscriptionCallback>,
+    ) -> Result<Self, GenericError> {
+        let (tx, mut rx): (_, mpsc::Receiver<FfiV2SubscribeRequest>) = mpsc::channel(10);
+
+        let handle = tokio::spawn(async move {
+            loop {
+                tokio::select! {
+                    item = subscription.next() => {
+                        match item {
+                            Some(Ok(envelope)) => callback.on_message(envelope.into()),
+                            Some(Err(e)) => log::error!("Stream error {}", e),
```

Comment:
Unlike the previous v2 implement, we don't return errors here. it's possible to pass them into the callback but Uniffi is finicky with passing `Result` types into the callback. V3 streaming also just logs the errors and moves on, so I did the same here